### DISCUSSION TITLE
GitVersion options with MS build

### DISCUSSION THREAD

**Discussion by Tognolo** [GitTools/GitVersion#2674] [Q&A] [answered]
Hello everybody,

I'm trying to migrate to GitVersion.MSBuild but I'm struggling with configuration. I've a GitVersion.yml configuration file and everything was fine, but for MSBuild seems not enough. My problem is that I've a project with a submodule, and GitVersion get a version for the main git project and a version for the submodule. Because of the commit reference to the submodule, I usually have a detached commit for this one, and here comes the troubles, because I haven't found how to set a fallback.

Digging into the code, I've found in `GitVersion.GitVersionContextFactory.Create` where is my problem:

```cs
public GitVersionContext Create(GitVersionOptions gitVersionOptions)
{
    var currentBranch = repositoryStore.GetTargetBranch(gitVersionOptions.RepositoryInfo.TargetBranch);
    if (currentBranch == null)
        throw new InvalidOperationException("Need a branch to operate on");

    var configuration = configProvider.Provide(overrideConfig: options.Value.ConfigInfo.OverrideConfig);

    var currentCommit = repositoryStore.GetCurrentCommit(currentBranch, gitVersionOptions.RepositoryInfo.CommitId);

    if (currentBranch.IsDetachedHead)
    {
        var branchForCommit = repositoryStore.GetBranchesContainingCommit(currentCommit, onlyTrackedBranches: gitVersionOptions.Settings.OnlyTrackedBranches).OnlyOrDefault();
        currentBranch = branchForCommit ?? currentBranch;
    }

    var currentBranchConfig = branchConfigurationCalculator.GetBranchConfiguration(currentBranch, currentCommit, configuration);
    var effectiveConfiguration = configuration.CalculateEffectiveConfiguration(currentBranchConfig);
    var currentCommitTaggedVersion = repositoryStore.GetCurrentCommitTaggedVersion(currentCommit, effectiveConfiguration);
    var numberOfUncommittedChanges = repositoryStore.GetNumberOfUncommittedChanges();

    return new GitVersionContext(currentBranch, currentCommit, configuration, effectiveConfiguration, currentCommitTaggedVersion, numberOfUncommittedChanges);
}
```

`currentBranch` can be `master` or `default` for me, but I'm unable to understand how to set `gitVersionOptions`, so it comes out as `No branch`. Because `currentBranch.IsDetachedHead` is true, code look for a branch for the commit, but again it looks for tracked and untracked branches, in my case I've `master` and `origin/master`, so `branchForCommit` is `null` and `currentBranch` is still `No branch`. After that, because the commit has no version tag, I've an exception in `GitVersion.VersionCalculation.NextVersionCalculator.EnsureHeadIsNotDetached`.

I can manually set the right branch, but I can't do it on Azure devops pipeline.

Any suggestions?

**Comment by asbjornu** [comment]
[As documented](https://gitversion.net/docs/reference/requirements), GitVersion does not support a detached `HEAD`. I'm not sure, but perhaps providing the environment variable `Git_Branch` helps.

**Reply by Tognolo** [reply]
Ok. Using submodules, detached `HEAD` is a common thing. Previous version of GitVersion support missing branch (at least can compile submodules code). I'm stuck with it.

I'd like to add fallback configuration to GitVersion.yml, because environment variable is a pain.

**Reply by asbjornu** [reply, marked as answer]
It's unfortunately not possible to configure the currently built branch with configuration, sorry. 😕
